File: w1-pseudorandom-number-generation/prng-uniform.R
mixed_congruential <- function(
    seed, n_samples, a = 7^5, m = (2 ^ 31) - 1, c=(2 ^ 10) - 1) {
  x <- rep(NA, times=n_samples)
  x[1] <- seed
  hasRepeat <- FALSE
  for (i in 2:n_samples) {
    val <- (x[i - 1] * a + c) %% m
    # If val is already present, the values will repeat
    if (val %in% x[1:i]) {
      hasRepeat <- TRUE
      break
    }
    x[i] <- val
  }
  if (hasRepeat) {  # loop broke because of repeating values, so fill the rest
    reps <- x[1:(i - 1)]  # the original sequence to be repeated
    x[i:n_samples] <- rep(reps, length.out=(n_samples - i + 1))
  }
  return(x / m)
}


uniform_dist <- function(seed, n_samples, a, b) {
  x <- mixed_congruential(seed, n_samples)
  return((b - a) * x + a)
}


# Usage: Draw 1k samples ~ U(3, 7)
y <- uniform_dist(0, 1000, 3, 7)
par(mfrow=c(1, 2))
hist(y, main="", xlab="", sub="Histogram", font.sub=1)
plot.ts(y, main="", xlab="", sub="Trace", font.sub=1)
mtext("Multiplicative Congruential Method: 1000 samples, seed = 0", outer=TRUE,
      side=3, line=-2, cex=1.5)

File: w1-pseudorandom-number-generation/prng_uniform.py
import numpy as np


def mixed_congruential(seed, n_samples, a=7**5,
                       m=(2**31 - 1), c=(2 ** 10 - 1)):
    """Uniform distribution ~U[0, 1] with the mixed congruential method."""
    x = np.empty(n_samples)
    x[0] = seed
    has_repeat = False
    for i in range(1, n_samples):
        val = (x[i - 1] * a + c) % m
        if val in x[:i]:
            has_repeat = True
            break
        x[i] = val
    if has_repeat:
        x = np.tile(x[:i], reps=(n_samples // i + 1,))[:n_samples]
    return x / m


def uniform_dist(seed, n_samples, A=0, B=1, **mc_params):  # NOQA: N803
    """A uniform distribution ~U[A, B]."""
    x = mixed_congruential(seed, n_samples, **mc_params)
    return (B - A) * x + A
